File: web/src/game/Room.ts
import { V2 } from "./V2";
import { Cube, Action, Weight } from "./Cube";
import { Field } from "./Field";
import { Landslip } from "./Landslip";
import { Controls } from "./Controls";
import { FinderAlg } from "./FinderAlg";
import { MouseControl } from "./MouseControl";
import { Unit, InputProvider } from "./Unit";

/**
 * One round of the puzzle: apply the previous round's pending moves, check
 * for deaths, let goal_escape models walk out through the border, then let
 * unsupported items fall. Port of legacy/src/level/Room.h/.cpp, reduced to
 * just the simulation - drawing, sound and save/undo are dropped; the
 * active-fish-switch scheme lives in Controls.ts (docs/016), mouse
 * pathfinding/pushing in MouseControl.ts/FinderAlg.ts (docs/017).
 */
export class Room {
  readonly field: Field;
  readonly models: Cube[] = [];
  private readonly controls = new Controls();
  private readonly mouseControl = new MouseControl(this.controls, new FinderAlg());
  private lastAction: Action = Action.NO;
  /** Cubes that died (isAlive -> false) during the most recently finished round. */
  lastDead: Cube[] = [];
  /** Weight of whatever just landed after falling this round (NONE if
   *  nothing did) - legacy's Landslip::getImpact(), read by LevelScene to
   *  play an impact sound (docs/018). Ported since docs/007 but never
   *  read until now. */
  lastImpact: Weight = Weight.NONE;
  /** legacy Room::m_fastFalling - when set, a round settles every pending fall
   *  at once instead of one step per round (windoze uses it while the player
   *  solves the bonus). Outcome-identical to normal falling, just faster. */
  private fastFalling = false;

  constructor(
    readonly w: number,
    readonly h: number,
  ) {
    this.field = new Field(w, h);
  }

  /** game_setFastFalling(value) via the live-Lua bridge (docs/035). */
  setFastFalling(value: boolean): void {
    this.fastFalling = value;
  }

  /** game_checkActive(): switch control away from an active fish that can no
   *  longer drive (e.g. just made busy) - legacy Room::checkActive/
   *  Controls::checkActive. See docs/035. */
  checkActive(): void {
    this.controls.checkActive();
  }

  addModel(model: Cube, unit?: Unit): number {
    model.rules.takeField(this.field);
    this.models.push(model);
    model.index = this.models.length - 1;
    if (unit) {
      this.controls.addUnit(unit);
    }
    return model.index;
  }

  isFresh(): boolean {
    return this.lastAction === Action.NO;
  }

  isFalling(): boolean {
    return this.lastAction === Action.FALL;
  }

  /** One full engine round: settle any pending move/fall, then (if settled)
   *  accept new input - keyboard first, mouse only if keyboard produced no
   *  move this round (matches the original's real precedence). */
  nextRound(input: InputProvider): void {
    // legacy Room::nextRound: fast-falling settles every pending fall in this
    // one round before accepting input, so the player never waits on the main
    // room settling while solving windoze's bonus (docs/035).
    if (this.fastFalling) {
      this.fastForwardSettle();
    } else {
      this.beginFall();
    }
    if (this.isFresh()) {
      if (this.driving(input)) {
        this.lastAction = Action.MOVE;
      } else if (this.mouseControl.mouseDrive(input)) {
        this.lastAction = Action.MOVE;
      }
    }
    this.updateMoveStreaks();
  }

  private updateMoveStreaks(): void {
    for (const m of this.models) m.rules.updateMoveStreak();
  }

  private beginFall(): void {
    this.prepareRound();
    this.lastAction = Action.NO;
    // Reset every round regardless of which branch below runs, so a round
    // where nothing falls (or fallout() pre-empts falldown() entirely)
    // correctly reports "no impact" rather than a stale prior value.
    this.lastImpact = Weight.NONE;

    if (this.fallout()) {
      this.lastAction = Action.MOVE;
    } else if (this.falldown()) {
      this.lastAction = Action.FALL;
    }
  }

  /** Apply last round's pending moves, then check who died as a result. */
  private prepareRound(): void {
    for (const m of this.models) m.rules.freeOldPos();
    for (const m of this.models) m.rules.occupyNewPos();

    this.lastDead = [];
    for (const m of this.models) {
      if (m.rules.checkDead(this.lastAction)) {
        this.lastDead.push(m);
      }
    }

    for (const m of this.models) m.rules.changeState();
  }

  /** Let goal_escape/goal_out models walk toward and through the border. */
  private fallout(): boolean {
    let wentOut = false;
    for (const m of this.models) {
      if (!m.isLost) {
        const outDepth = m.rules.actionOut();
        if (outDepth > 0) wentOut = true;
      }
    }
    return wentOut;
  }

  private falldown(): boolean {
    const slip = new Landslip(this.models);
    const falling = slip.computeFall();
    this.lastImpact = slip.getImpact();
    return falling;
  }

  private driving(input: InputProvider): boolean {
    return this.controls.driving(input);
  }

  /** Space key: switch to the next drivable fish - legacy's Room::switchFish(). */
  switchFish(): void {
    this.controls.requestSwitch();
  }

  /** Returns the model occupying `loc`, or the shared border Cube for an
   *  out-of-bounds location - legacy's Room::askField(). */
  askField(loc: V2): Cube | null {
    return this.field.getModel(loc);
  }

  /** Click-to-select - legacy's Room::controlMouse() (left-button branch). */
  selectFish(model: Cube): void {
    this.controls.activateSelected(model);
  }

  /** Every move symbol recorded so far, in order - legacy's Room::
   *  stepCounter(). See docs/021. */
  getMoves(): string {
    return this.controls.getMoves();
  }

  getStepCount(): number {
    return this.controls.getStepCount();
  }

  /** The currently-active fish (arrows drive it) - legacy's Controls::m_active.
   *  Drives the shared animation clock's phase count (docs/046). */
  getActiveUnit(): Unit | null {
    return this.controls.getActive();
  }

  /** Fast, render-free replay of one recorded move symbol - legacy's
   *  Room::loadMove(): settle any pending falls first, then apply exactly
   *  this one move. The move's own consequences (position committing, and
   *  whatever it triggers falling) are picked up by the *next* loadMove()/
   *  settleAll() call's own settling, exactly like the interactive round
   *  pipeline's decide-this-round/apply-next-round split (docs/007) - just
   *  run back to back with no real-time pause instead of one real round at
   *  a time. Used by the headless solution validator (docs/022).
   *  @throws Error if `symbol` doesn't name a valid move for any unit here. */
  loadMove(symbol: string): void {
    this.fastForwardSettle();
    if (!this.controls.makeMove(symbol)) {
      throw new Error(`invalid move: "${symbol}"`);
    }
    this.lastAction = Action.MOVE;
    this.updateMoveStreaks();
  }

  /** Settles any pending falls with nothing left to drive - call once
   *  after the last loadMove() in a solution to apply its consequences
   *  before checking isSolved(). */
  settleAll(): void {
    this.fastForwardSettle();
  }

  /** One round of *watchable* replay - same physics shape as nextRound()
   *  (settle, then only accept new input once fresh), but sourced from a
   *  recorded move string instead of live keyboard/mouse. Unlike
   *  loadMove(), does not fast-forward through pending falls - each call
   *  is exactly one round, so a caller ticking this on a real-time timer
   *  sees the same falling/sliding animation live play would. Doesn't
   *  consume `symbol` unless the round is actually fresh, matching how it
   *  was originally recorded (docs/021: a symbol is only ever captured
   *  when isFresh()). See docs/025.
   *  @return whether `symbol` was consumed this round. */
  replayRound(symbol: string | null): boolean {
    if (this.fastFalling) {
      this.fastForwardSettle();
    } else {
      this.beginFall();
    }
    let consumed = false;
    if (this.isFresh() && symbol !== null) {
      consumed = this.controls.makeMove(symbol);
      if (consumed) this.lastAction = Action.MOVE;
    }
    this.updateMoveStreaks();
    return consumed;
  }

  /** Start a show-driven round - legacy's Level::nextShowAction()'s
   *  room->beginFall(): settle pending falls before the show command runs.
   *  Public entry for the briefcase auto-play "show" (docs/031, Phase 2),
   *  which drives the round itself (input disabled) instead of nextRound(). */
  beginShowRound(): void {
    this.beginFall();
  }

  /** Apply one scripted "show" move - legacy's Room::makeMove(): only lands
   *  when the room is fresh (else the show command retries next round), throws
   *  on a fresh-but-impossible move (caught by the show driver as a graceful
   *  end). Assumes beginShowRound() already ran this round. See docs/031. */
  showMove(symbol: string): boolean {
    if (!this.isFresh()) return false;
    if (!this.controls.makeMove(symbol)) {
      throw new Error(`show move not possible: "${symbol}"`);
    }
    this.lastAction = Action.MOVE;
    this.updateMoveStreaks();
    return true;
  }

  /** Repeatedly resolves falls/fallout until nothing is left pending
   *  (isFresh()) - legacy's loadMove()'s "let object to fall fast" loop.
   *  Bounded defensively: a real level can never fall forever, so hitting
   *  the cap means something is wrong, not just a long level. */
  private fastForwardSettle(): void {
    const MAX_SETTLE_ROUNDS = 1000;
    let rounds = 0;
    do {
      this.beginFall();
      if (++rounds > MAX_SETTLE_ROUNDS) {
        throw new Error("settling did not converge - possible infinite fall loop");
      }
    } while (!this.isFresh());
  }

  /** No unit will ever be able to move again (all driven fish dead/lost). */
  cannotMove(): boolean {
    return this.controls.cannotMove();
  }

  /** All goals can still possibly be satisfied (false forever once a goal_escape fish dies). */
  isSolvable(): boolean {
    return this.models.every((m) => !m.isWrong);
  }

  /** All goals are currently satisfied and nothing is still moving/falling. */
  isSolved(): boolean {
    return this.isFresh() && this.models.every((m) => m.isSatisfy);
  }
}

File: solver/src/FishFillets.Physics/Room.cs
using System.Runtime.CompilerServices;

namespace FishFillets.Physics;

/// <summary>
/// One playable room: the occupancy grid plus every model's live state, and the
/// round pipeline that drives them. Port of legacy/src/level/Room.h/.cpp merged
/// with Field.cpp and MarkMask.cpp (web/src/game/{Room,Field,MarkMask}.ts) - the
/// rules themselves live in Room.Rules.cs, gravity in Room.Landslip.cs, and
/// move symbols in Room.Controls.cs, all partials of this class.
///
/// Merged rather than mirrored as separate classes on purpose: Field/MarkMask
/// are per-model objects in the original that exist only to reach the one shared
/// grid, and collapsing them removes an indirection from the hottest inner loop
/// (getResist) without changing any behaviour.
///
/// Dropped vs. the browser port, because the solver has no player and no
/// screen: input handling (InputProvider/Controls.driving/MouseControl/
/// FinderAlg), the active-fish switching scheme, animation state and the
/// visual move-streak. What remains is exactly the simulation - moving,
/// pushing, falling, dying, escaping - plus the win/lose predicates.
/// See solver/docs/001.
/// </summary>
public sealed partial class Room
{
    /// <summary>Empty water. Stored in the grid; never a model index.</summary>
    public const int Empty = -1;

    /// <summary>
    /// Rounds a dead model stays masked and solid before it is unmasked and
    /// removed - port of EffectDisintegrate's pixel-dissolve counter
    /// (ceil(400/30) = 14 render frames, counted here in physics rounds).
    /// See docs/011.
    ///
    /// A solver normally prunes the moment a fish dies (both fish goals fail on
    /// death, so <see cref="IsSolvable"/> goes false), which makes this
    /// unreachable in practice - it is ported anyway so that any state the
    /// search does keep behaves exactly like the game.
    /// </summary>
    private const byte DeathRemoveRounds = 14;

    private const int MaxSettleRounds = 1000;

    public Level Level { get; }

    private readonly short[] _field;
    private readonly ModelState[] _models;
    private readonly ResistArena _arena;
    private readonly ModelCollector _pads;
    private readonly ModelCollector _falling;
    private readonly ModelCollector _heavier;
    private readonly bool[] _stoned;

    private readonly int _width;
    private readonly int _height;
    private readonly int _borderIndex;

    private Action _lastAction = Action.No;
    private Weight _lastImpact = Weight.None;
    private bool _fastFalling;
    private int _stepCount;
    private int _lastDeadCount;

    public Room(Level level)
    {
        Level = level;
        _width = level.Width;
        _height = level.Height;
        _borderIndex = level.BorderIndex;

        _field = new short[_width * _height];
        _field.AsSpan().Fill(Empty);

        int slots = level.Models.Length;
        _models = new ModelState[slots];
        _arena = new ResistArena(slots, Math.Max(_width, _height));
        _pads = new ModelCollector(slots);
        _falling = new ModelCollector(slots);
        _heavier = new ModelCollector(slots);
        _stoned = new bool[slots];

        Reset();
    }

    /// <summary>Number of real models (the border excluded) - loop bound everywhere.</summary>
    public int ModelCount => Level.ModelCount;

    public ref readonly ModelState State(int model) => ref _models[model];

    /// <summary>Nothing is still moving or falling (legacy Room::isFresh()).</summary>
    public bool IsFresh => _lastAction == Action.No;

    /// <summary>Moves successfully applied so far (legacy StepCounter).</summary>
    public int StepCount => _stepCount;

    /// <summary>Models that died during the most recently finished round.</summary>
    public int LastDeadCount => _lastDeadCount;

    /// <summary>Weight of whatever just landed after falling (legacy Landslip::getImpact()).</summary>
    public Weight LastImpact => _lastImpact;

    /// <summary>
    /// Restores the room to its as-loaded state. Cheaper than building a new
    /// <see cref="Room"/> (no shape re-parse, no reallocation), so a search or a
    /// batch verify reuses one instance per thread.
    /// </summary>
    public void Reset()
    {
        _field.AsSpan().Fill(Empty);
        _lastAction = Action.No;
        _lastImpact = Weight.None;
        _fastFalling = false;
        _stepCount = 0;
        _lastDeadCount = 0;

        for (int i = 0; i < _models.Length; i++)
        {
            ModelDef def = Level.Models[i];
            _models[i] = new ModelState
            {
                X = def.X,
                Y = def.Y,
                Weight = def.Weight,
                IsAlive = def.IsAlive,
                IsLeft = def.IsLeft,
                OutDir = def.OutDir,
                OutCapacity = def.OutCapacity,
                Dir = Dir.No,
                TouchDir = Dir.No,
            };
        }

        // legacy Room::addModel -> Rules::takeField, in declaration order: model
        // 0 (the room's wall shape) masks first, then the items on top of it.
        // The border is never masked - it sits out of bounds at (-1,-1), where
        // writes are dropped anyway.
        for (int i = 0; i < ModelCount; i++)
        {
            if (HasPlacedResist(i, _models[i].X, _models[i].Y))
            {
                throw new InvalidDataException(
                    $"{Level.Name}: position is occupied at build: model {i} ({Level.Models[i].Kind}) " +
                    $"at ({_models[i].X},{_models[i].Y})");
            }

            Mask(i);
        }

        // A room's items can start unsupported - the original settles them at
        // load, before the player gets control (docs/019's "initial level-load
        // input freeze"). Doing it here is what lets ApplyMove() settle *after*
        // its move and still match the original's settle-before-move order.
        SettleAll();
    }

    // ---------------------------------------------------------------- Field --

    /// <summary>
    /// The model occupying (x, y): a model index, <see cref="Empty"/> for empty
    /// in-bounds water, or the border index for anything out of bounds. Port of
    /// Field::getModel() - the original's "hack border everywhere in outer
    /// space".
    /// </summary>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public int GetModel(int x, int y) =>
        (uint)x < (uint)_width && (uint)y < (uint)_height
            ? _field[(y * _width) + x]
            : _borderIndex;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private void SetModel(int x, int y, short model)
    {
        if ((uint)x < (uint)_width && (uint)y < (uint)_height)
        {
            _field[(y * _width) + x] = model;
        }
    }

    /// <summary>Clears a cell, but only if this model is the one occupying it (Field::setModel's toOverride).</summary>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private void ClearModel(int x, int y, int model)
    {
        if ((uint)x < (uint)_width && (uint)y < (uint)_height)
        {
            ref short cell = ref _field[(y * _width) + x];
            if (cell == model)
            {
                cell = Empty;
            }
        }
    }

    // ------------------------------------------------------------- MarkMask --

    /// <summary>Writes this model's shape into the grid (MarkMask::mask()).</summary>
    private void Mask(int model)
    {
        Shape shape = Level.Models[model].Shape;
        int x = _models[model].X, y = _models[model].Y;
        for (int m = 0; m < shape.MarkX.Length; m++)
        {
            SetModel(x + shape.MarkX[m], y + shape.MarkY[m], (short)model);
        }
    }

    /// <summary>Lifts this model out of the grid (MarkMask::unmask()).</summary>
    private void Unmask(int model)
    {
        Shape shape = Level.Models[model].Shape;
        int x = _models[model].X, y = _models[model].Y;
        for (int m = 0; m < shape.MarkX.Length; m++)
        {
            ClearModel(x + shape.MarkX[m], y + shape.MarkY[m], model);
        }
    }

    /// <summary>
    /// Everything (other than this model) occupying the cells its shape would
    /// cover one step along <paramref name="dir"/>. Port of
    /// MarkMask::getResist(); the returned frame must be consumed inside its
    /// `using` - see <see cref="ResistArena"/>.
    /// </summary>
    private ResistArena.Frame GetResist(int model, Dir dir) =>
        GetPlacedResist(model, _models[model].X + dir.X(), _models[model].Y + dir.Y());

    /// <summary>Everything (other than this model) occupying the cells its shape would cover at (lx, ly).</summary>
    private ResistArena.Frame GetPlacedResist(int model, int lx, int ly)
    {
        Shape shape = Level.Models[model].Shape;
        ResistArena.Frame frame = _arena.Open();
        for (int m = 0; m < shape.MarkX.Length; m++)
        {
            int resist = GetModel(lx + shape.MarkX[m], ly + shape.MarkY[m]);
            if (resist != Empty && resist != model)
            {
                frame.Push(resist);
            }
        }

        return frame;
    }

    /// <summary>Allocation-free "is anything in the way at (lx, ly)" - avoids opening a frame just to test emptiness.</summary>
    private bool HasPlacedResist(int model, int lx, int ly)
    {
        Shape shape = Level.Models[model].Shape;
        for (int m = 0; m < shape.MarkX.Length; m++)
        {
            int resist = GetModel(lx + shape.MarkX[m], ly + shape.MarkY[m]);
            if (resist != Empty && resist != model)
            {
                return true;
            }
        }

        return false;
    }

    /// <summary>A direction leading out of the room with nothing blocking, or <see cref="Dir.No"/> (MarkMask::getBorderDir()).</summary>
    private Dir GetBorderDir(int model)
    {
        Shape shape = Level.Models[model].Shape;
        int x = _models[model].X, y = _models[model].Y;
        for (int m = 0; m < shape.MarkX.Length; m++)
        {
            int cx = x + shape.MarkX[m], cy = y + shape.MarkY[m];
            if (cx == 0 && CanMoveOthers(model, Dir.Left, Weight.Fixed)) return Dir.Left;
            if (cx == _width - 1 && CanMoveOthers(model, Dir.Right, Weight.Fixed)) return Dir.Right;
            if (cy == 0 && CanMoveOthers(model, Dir.Up, Weight.Fixed)) return Dir.Up;
            if (cy == _height - 1 && CanMoveOthers(model, Dir.Down, Weight.Fixed)) return Dir.Down;
        }

        return Dir.No;
    }

    /// <summary>Every cell of this model's shape is out of the room (MarkMask::isFullyOut()).</summary>
    private bool IsFullyOut(int model)
    {
        Shape shape = Level.Models[model].Shape;
        int x = _models[model].X, y = _models[model].Y;
        for (int m = 0; m < shape.MarkX.Length; m++)
        {
            if (GetModel(x + shape.MarkX[m], y + shape.MarkY[m]) != _borderIndex)
            {
                return false;
            }
        }

        return true;
    }

    // --------------------------------------------------------- Cube helpers --

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private bool IsWall(int model) => _models[model].Weight >= Weight.Fixed;

    /// <summary>How much this model can push/hold. Immutable, so it lives on <see cref="ModelDef"/>.</summary>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private Weight Power(int model) => Level.Models[model].Power;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private bool IsBorder(int model) => model == _borderIndex;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private bool ShouldGoOut(int model)
    {
        GoalKind goal = Level.Models[model].Goal;
        return goal is GoalKind.Out or GoalKind.Escape;
    }

    /// <summary>Port of Goal::isSatisfy() - see <see cref="GoalKind"/> for the tri-state mapping.</summary>
    private bool IsSatisfy(int model)
    {
        ref ModelState m = ref _models[model];
        return Level.Models[model].Goal switch
        {
            GoalKind.No => true,
            GoalKind.Out => m.IsOut,
            GoalKind.Escape => m.IsOut && m.IsAlive,
            GoalKind.Alive => m.IsAlive,
            _ => true,
        };
    }

    /// <summary>Port of Goal::isWrong() - this goal can never be satisfied again.</summary>
    private bool IsWrong(int model)
    {
        GoalKind goal = Level.Models[model].Goal;
        return (goal is GoalKind.Escape or GoalKind.Alive) && !_models[model].IsAlive;
    }

    private void ChangeDie(int model) => _models[model].IsAlive = false;

    /// <summary>Going out of the room - immovable while crossing the border (Cube::changeGoingOut()).</summary>
    private void ChangeGoingOut(int model) => _models[model].Weight = Weight.Fixed;

    private void ChangeGoOut(int model)
    {
        _models[model].IsOut = true;
        ChangeRemove(model);
    }

    private void ChangeRemove(int model)
    {
        ref ModelState m = ref _models[model];
        m.IsLost = true;
        m.Weight = Weight.None;
        m.X = -1000;
        m.Y = -1000;
    }

    /// <summary>
    /// A fish just went out through this plug (Cube::decOutCapacity()). Once
    /// spent, the plug stops being an output and becomes a normal light item.
    /// </summary>
    private void DecOutCapacity(int model)
    {
        ref ModelState m = ref _models[model];
        if (m.OutCapacity > 0)
        {
            m.OutCapacity--;
            if (m.OutCapacity == 0)
            {
                m.OutDir = Dir.No;
                m.Weight = Weight.Light;
                m.OutCapacity = -1;
            }
        }
    }

    // ------------------------------------------------------------- Outcomes --

    /// <summary>Every goal is currently satisfied and nothing is still moving (Room::isSolved()).</summary>
    public bool IsSolved()
    {
        if (!IsFresh)
        {
            return false;
        }

        for (int i = 0; i < ModelCount; i++)
        {
            if (!IsSatisfy(i))
            {
                return false;
            }
        }

        return true;
    }

    /// <summary>
    /// Every goal can still possibly be satisfied (Room::isSolvable()). Goes
    /// false forever the moment a fish dies - both fish goals require alive - so
    /// this is the search's primary prune.
    /// </summary>
    public bool IsSolvable()
    {
        for (int i = 0; i < ModelCount; i++)
        {
            if (IsWrong(i))
            {
                return false;
            }
        }

        return true;
    }

    /// <summary>No fish will ever move again - all dead or already out (Room::cannotMove()).</summary>
    public bool CannotMove()
    {
        foreach (UnitDef unit in Level.Units)
        {
            if (WillMove(unit.Model))
            {
                return false;
            }
        }

        return true;
    }

    /// <summary>game_setFastFalling(): settle every pending fall in one round (windoze, docs/035).</summary>
    public void SetFastFalling(bool value) => _fastFalling = value;

    /// <summary>model_setBusy(): freeze a fish out of player control (windoze, docs/035).</summary>
    public void SetBusy(int model, bool busy) => _models[model].Busy = busy;
}
